### PR TITLE
Add support for Samsung browsers

### PR DESCRIPTION
**Pull request by julian-lai**
Currently, `UserAgent` reports Samsung browsers as Chrome.

### CODE REVIEW

File: lib/user_agent/browsers/samsung.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
+class UserAgent
+  module Browsers
+    class Samsung < Base
+
+      SAMSUNG_BROWSER = 'SamsungBrowser'
+
+      def self.extend?(agent)
+        agent.detect { |useragent| useragent.product == SAMSUNG_BROWSER}
```
Comment (ziggythehamster) [inline lib/user_agent/browsers/samsung.rb]:
```suggestion
        agent.detect { |useragent| useragent.product == SAMSUNG_BROWSER }
```

File: lib/user_agent/browsers/samsung.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      end
+
+      def version
+        self.detect { |useragent| useragent.product == SAMSUNG_BROWSER}.version
```
Comment (ziggythehamster) [inline lib/user_agent/browsers/samsung.rb]:
```suggestion
        self.detect { |useragent| useragent.product == SAMSUNG_BROWSER }.version
```

File: lib/user_agent/browsers/samsung.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+      ##
+      # @return [Boolean]
+      #     SamsungBrowser is a mobile device
+      def mobile?
+        true
+      end
+
+      def browser
+        SAMSUNG_BROWSER
```
Comment (ziggythehamster) [inline lib/user_agent/browsers/samsung.rb]:
```suggestion
        'Samsung Internet Browser'
```

This seems to be the name that Samsung prefers (according to here: https://play.google.com/store/apps/details?id=com.sec.android.app.sbrowser)

File: lib/user_agent/browsers/samsung.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
+class UserAgent
+  module Browsers
+    class Samsung < Base
+
```
Comment (ziggythehamster) [inline lib/user_agent/browsers/samsung.rb]:
```suggestion
    class Samsung < WebKit
```

File: lib/user_agent/browsers/samsung.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        SAMSUNG_BROWSER
+      end
+
+      def platform
+        ANDROID
+      end
+
+      def os
+        ANDROID
+      end
```
Comment (ziggythehamster) [inline lib/user_agent/browsers/samsung.rb]:
```suggestion
```

`WebKit` will auto-fill these correctly (i.e., `os` will be `Android 11` in your spec)

File: spec/browsers/samsung_user_agent_spec.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  it { expect(useragent).not_to be_desktop }
+  it { expect(useragent).not_to be_speaker }
+  it { expect(useragent).not_to be_bot }
+end
+
+describe "UserAgent: Mozilla/5.0 (Linux; Android 11; SAMSUNG SM-G991B) AppleWebKit/537.36 (KHTML, like Gecko) SamsungBrowser/11.1 Chrome/75.0.3770.143 Mobile Safari/537.36" do
+  let(:useragent) { UserAgent.parse("Mozilla/5.0 (Linux; Android 11; SAMSUNG SM-G991B) AppleWebKit/537.36 (KHTML, like Gecko) SamsungBrowser/11.1 Chrome/75.0.3770.143 Mobile Safari/537.36") }
+
+  it_behaves_like 'SamsungBrowser', '11.1', 'Android', 'Android'
+end
```
Comment (ziggythehamster) [inline spec/browsers/samsung_user_agent_spec.rb]:
```suggestion
end

```

For reasons I don't entirely understand, Git patches don't apply cleanly in some situations when there's not a newline at the end of the file. :)

File: spec/browsers/samsung_user_agent_spec.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  it "returns '#{version}' as its version" do
+    expect(useragent.version).to eq(version)
+  end
+
+  it "returns 'Android' as its platform" do
+    expect(useragent.platform).to eq('Android')
+  end
+
+  it "returns 'Android' as its operating system" do
+    expect(useragent.os).to eq('Android')
```
Comment (ziggythehamster) [inline spec/browsers/samsung_user_agent_spec.rb]:
```suggestion
    expect(useragent.os).to eq(os)
```

File: spec/browsers/samsung_user_agent_spec.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+  it "returns 'SamsungBrowser' as its browser" do
+    expect(useragent.browser).to eq('SamsungBrowser')
+  end
+
+  it "returns '#{version}' as its version" do
+    expect(useragent.version).to eq(version)
+  end
+
+  it "returns 'Android' as its platform" do
+    expect(useragent.platform).to eq('Android')
```
Comment (ziggythehamster) [inline spec/browsers/samsung_user_agent_spec.rb]:
```suggestion
    expect(useragent.platform).to eq(platform)
```

File: spec/browsers/samsung_user_agent_spec.rb [hunk 9]
```diff
@@ -0,0 +1,30 @@
+require 'user_agent'
+
+shared_examples 'SamsungBrowser' do |version, platform, os|
+  it "returns 'SamsungBrowser' as its browser" do
+    expect(useragent.browser).to eq('SamsungBrowser')
```
Comment (ziggythehamster) [inline spec/browsers/samsung_user_agent_spec.rb]:
```suggestion
    expect(useragent.browser).to eq('Samsung Internet Browser')
```

File: spec/browsers/samsung_user_agent_spec.rb [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  end
+
+  it "returns '#{version}' as its version" do
+    expect(useragent.version).to eq(version)
+  end
+
+  it "returns 'Android' as its platform" do
```
Comment (ziggythehamster) [inline spec/browsers/samsung_user_agent_spec.rb]:
```suggestion
  it "returns '#{platform}' as its platform" do
```

File: spec/browsers/samsung_user_agent_spec.rb [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    expect(useragent.platform).to eq('Android')
+  end
+
+  it "returns 'Android' as its operating system" do
```
Comment (ziggythehamster) [inline spec/browsers/samsung_user_agent_spec.rb]:
```suggestion
  it "returns '#{os}' as its operating system" do
```

File: spec/browsers/samsung_user_agent_spec.rb [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+    expect(useragent.os).to eq('Android')
+  end
+
+  it { expect(useragent).to be_mobile }
+  it { expect(useragent).not_to be_desktop }
+  it { expect(useragent).not_to be_speaker }
+  it { expect(useragent).not_to be_bot }
+end
+
+describe "UserAgent: Mozilla/5.0 (Linux; Android 11; SAMSUNG SM-G991B) AppleWebKit/537.36 (KHTML, like Gecko) SamsungBrowser/11.1 Chrome/75.0.3770.143 Mobile Safari/537.36" do
```
Comment (ziggythehamster) [inline spec/browsers/samsung_user_agent_spec.rb:26]: Do you have any other UAs you can add to the spec? I'll see if I have any I can give you

File: spec/browsers/samsung_user_agent_spec.rb [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+  let(:useragent) { UserAgent.parse("Mozilla/5.0 (Linux; Android 11; SAMSUNG SM-G991B) AppleWebKit/537.36 (KHTML, like Gecko) SamsungBrowser/11.1 Chrome/75.0.3770.143 Mobile Safari/537.36") }
+
+  it_behaves_like 'SamsungBrowser', '11.1', 'Android', 'Android'
```
Comment (ziggythehamster) [inline spec/browsers/samsung_user_agent_spec.rb]:
```suggestion
  it_behaves_like 'SamsungBrowser', '11.1', 'Android', 'Android 11'
```

**Review by ziggythehamster — CHANGES_REQUESTED** [review]
Thanks for the contribution! I'll see if I can find more UAs to test.

**Comment by ziggythehamster** [comment]
Here's some more user agents to add to your spec:

```
Dalvik/2.1.0 (Linux; U; Android 10; SM-N975U Build/QP1A.190711.020) AppleWebKit/537.36 (KHTML, like Gecko) SamsungBrowser/4.0 Chrome/44.0.2403.133 Mobile Safari/537.36
Mozilla/5.0 (Linux; Android 10; 9032Z) AppleWebKit/537.36 (KHTML, like Gecko) SamsungBrowser/13.2 Chrome/83.0.4103.106 Mobile Safari/537.36
Firefox/5.0 (Linux; Android 5.0.1; SAMSUNG GT-I9505 Build/LRX22C) AppleWebKit/537.36 (KHTML, like Gecko) SamsungBrowser/2.1 Chrome/34.0.1847.76 Mobile Safari/537.36
Mozilla/5.0 (Linux; Android 10.0; SAMSUNG SM-G960U Build/QP1A.190711.020) AppleWebKit/537.36 (KHTML, like Gecko) SamsungBrowser/5.2 Chrome/51.0.2704.106 Mobile VR Safari/537.36
Mozilla/5.0 (Linux; Android 4.4.4; SAMSUNG SM-G318H/G318HJVS0AQC2 Build/KTU84P) AppleWebKit/537.36 (KHTML, like Gecko) SamsungBrowser/2.0 Chrome/34.0.1847.76 Mobile Safari/537.36
```

That Firefox one is funny, because it's obviously still Chrome/WebKit based, they just decided to spoof the user-agent in one version.

**Comment by julian-lai** [comment]
Hey @ziggythehamster, thank you for all the feedback and the additional UA examples! I've applied to changes to my PR :)

**Review by ziggythehamster — APPROVED** [review]

**Pull request merged**